bin/check-es-heap.rb

```diff
@@ -120,46 +126,73 @@ def acquire_es_resource(resource)
     warning 'Elasticsearch API returned invalid JSON'
   end
 
-  def acquire_heap_data(return_max = false)
-    stats = if Gem::Version.new(acquire_es_version) >= Gem::Version.new('1.0.0')
-              acquire_es_resource('/_nodes/_local/stats')
-            else
-              acquire_es_resource('/_cluster/nodes/_local/stats')
-            end
-    node = stats['nodes'].keys.first
-    begin
-      if return_max
-        return stats['nodes'][node]['jvm']['mem']['heap_used_in_bytes'], stats['nodes'][node]['jvm']['mem']['heap_max_in_bytes']
+  def acquire_stats
+    if Gem::Version.new(acquire_es_version) >= Gem::Version.new('1.0.0')
+      if config[:all]
+        acquire_es_resource('/_nodes/stats')
       else
-        stats['nodes'][node]['jvm']['mem']['heap_used_in_bytes']
+        acquire_es_resource('/_nodes/_local/stats')
       end
-    rescue
-      warning 'Failed to obtain heap used in bytes'
+    elsif config[:all]
+      acquire_es_resource('/_cluster/nodes/stats')
+    else
+      acquire_es_resource('/_cluster/nodes/_local/stats')
     end
   end
 
-  def run
+  def acquire_heap_data(node)
+    return node['jvm']['mem']['heap_used_in_bytes'], node['jvm']['mem']['heap_max_in_bytes']
+  rescue
+    warning 'Failed to obtain heap used in bytes'
+  end
+
+  def acquire_heap_usage(heap_used, heap_max, node_name)
     if config[:percentage]
-      heap_used, heap_max = acquire_heap_data(true)
-      heap_used_ratio = ((100 * heap_used) / heap_max).to_i
-      message "Heap used in bytes #{heap_used} (#{heap_used_ratio}% full)"
-      if heap_used_ratio >= config[:crit]
-        critical
-      elsif heap_used_ratio >= config[:warn]
-        warning
-      else
-        ok
-      end
+      heap_usage = ((100 * heap_used) / heap_max).to_i
+      output = if config[:all]
+                 "Node #{node_name}: Heap used in bytes #{heap_used} (#{heap_usage}% full)\n"
+               else
+                 "Heap used in bytes #{heap_used} (#{heap_usage}% full)"
+               end
     else
-      heap_used = acquire_heap_data(false)
-      message "Heap used in bytes #{heap_used}"
-      if heap_used >= config[:crit]
-        critical
-      elsif heap_used >= config[:warn]
-        warning
-      else
-        ok
+      heap_usage = heap_used
+      output = config[:all] ? "Node #{node_name}: Heap used in bytes #{heap_used}\n" : "Heap used in bytes #{heap_used}"
+    end
+    [heap_usage, output]
+  end
+
+  def run
+    stats = acquire_stats
+    status_w = false
```

Comment:
rather than keeping track of the status and the message if you say use an array they can be the same thing. You could then use something like `warns.any?` to check if there are any warnings.